### DISCUSSION TITLE
Qdrant source code about upsert

### DISCUSSION THREAD

**Discussion by wdywjm** [qdrant/qdrant#2964] [General]
Hello, I'm interesting in Qdrant project, so I try to learn the source code. I want to figure out what's going on when some points insert/upsert into a collection, specially for a Hnsw index. But only what I can found is something for search or build whole index (segment optimizer workers do this), can not found about the process of upsert points into Hnsw index. Dose it never happen when upsert points or there is another stuff to update Hnsw index? A lot thanks if some favor.

**Comment by generall** [comment]
Hi @wdywjm, we don't upsert directly into HNSW. Instead, HNSW index is build during optimization step. In this way we can generate immutable data structures which are much more efficient than mutable index

**Reply by wdywjm** [reply]
thanks for your reply. Every insert/upsert will trigger the optimization? or optimization step will only happen as optimization configure,. but if the latter，this process is async and behind, we can not search the insert vector data immediately through HNSW until next optimization step? Beginner of rust and qdrant is awesome !

**Reply by generall** [reply]
> but if the latter，this process is async and behind, we can not search the insert vector data immediately through HNSW until next optimization step

All vectors are available for the search immediately after insertion, but after indexation is complete it will just search faster

**Reply by wdywjm** [reply]
Got it, appreciate a lot for your help！Looking forward to make contributions to Qdrant.
